transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,356 @@
+        let mut channel = self.channel.lock().map_err(|_| SenderError::Lock)?;
+        if channel.receiver_active {
+            let random_delay = crate::delay::random_delay();
+
+            trace!(delay_in_ms = random_delay.as_millis(), "generated mixer delay",);
+
+            let delayed_data: DelayedData<T> = (std::time::SystemTime::now() + random_delay, item).into();
+            channel.buffer.push(Reverse(delayed_data));
+
+            if let Some(waker) = channel.waker.as_ref() {
+                waker.wake_by_ref();
+            }
+
+            #[cfg(all(feature = "prometheus", not(test)))]
+            {
+                METRIC_QUEUE_SIZE.increment(1.0f64);
+
+                let weight = 1.0f64 / METRIC_DELAY_WINDOW as f64;
+                METRIC_MIXER_AVERAGE_DELAY.set(
+                    (weight * random_delay.as_millis() as f64) + ((1.0f64 - weight) * METRIC_MIXER_AVERAGE_DELAY.get()),
+                );
+            }
+
+            Ok(())
+        } else {
+            Err(SenderError::Closed)
+        }
+    }
+
+    fn poll_flush(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // The channel can only be closed by the receiver. The sender can be dropped at any point.
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Error returned by the [`Receiver`].
+#[derive(Debug, thiserror::Error)]
+pub enum ReceiverError {
+    /// The channel is closed due to receiver being dropped.
+    #[error("Channel is closed")]
+    Closed,
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Receiver object interacting with the mixer channel.
+///
+/// The receiver receives already mixed elements without any knowledge of
+/// the original order.
+pub struct Receiver<T> {
+    channel: Arc<Mutex<Channel<T>>>,
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+
+    fn poll_next(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut channel = self.channel.lock().unwrap();
+        let now = std::time::SystemTime::now();
+        if channel.sender_count > 0 {
+            if channel.buffer.peek().map(|x| x.0.release_at < now).unwrap_or(false) {
+                let data = channel
+                    .buffer
+                    .pop()
+                    .expect("The value should be present within the same locked access")
+                    .0
+                    .item;
+
+                trace!("yielding a value");
+
+                #[cfg(all(feature = "prometheus", not(test)))]
+                METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                return Poll::Ready(Some(data));
+            }
+
+            let waker = cx.waker().clone();
+            channel.waker = Some(waker);
```

Comment:
Shouldn't the waker be set only in cases where you're about to return `Poll::Pending` ?